### PR TITLE
OAuth service/refactor

### PR DESCRIPTION
**Pull request by jeremyruppel**
This creates an OAuth "service" class representing the https://www.flickr.com/services/oauth API methods. The class also has methods for signing OAuth requests, so we can just use that under the hood in our OAuth superagent plugin.

### CODE REVIEW

File: services/oauth.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+/**
+ * Creates an object with the standard OAuth 1.0 query params
+ * for this instance.
+ * @returns {Object}
+ */
+
+OAuth.prototype.params = function () {
+	return {
+		oauth_nonce: nonce(Date.now()),
+		oauth_timestamp: Date.now(), // FIXME shouldn't this be seconds instead of milliseconds?
```
Comment (pdokas) [inline services/oauth.js]: Does… does it not work if you pass the timestamp as seconds?

File: services/oauth.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * @returns String.
+ * @see https://oauth.net/core/1.0a/#sig_norm_param
+ * @private
+ */
+
+function query(req) {
+	var pairs = req.qsRaw.slice();
+
+	Object.keys(req.qs).forEach(function (key) {
+		pairs.push(key + '=' + encodeURIComponent(req.qs[key]));
```
Comment (alex-seville) [inline services/oauth.js]: does req.qs already normalize objects to strings/numeric, or do we need to JSON.stringify objects here too?

File: services/oauth.js (hunk 1 above)
Comment (jeremyruppel) [inline services/oauth.js]: Actually I just noticed this FIXME as well and it does work using seconds. Fixing that now.

File: test/services.oauth.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
+var OAuth = require('../services/oauth');
+var request = require('superagent');
+var assert = require('assert');
+var time = require('timemachine');
```
Comment (alex-seville) [inline test/services.oauth.js]: too good for my time mocking lib, I see how it is

File: test/services.oauth.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+				oauth_token: 'token',
+				oauth_verifier: 'verfier',
+				oauth_nonce: '0d6e24f8073fe797e4d0187b95a141543e43698d',
+				oauth_timestamp: 499162800000,
+				oauth_consumer_key: subject.consumerKey,
+				oauth_signature_method: 'HMAC-SHA1',
+				oauth_version: '1.0',
+				oauth_signature: 'Qyhwc9332I2rv4OT/ecfI6yNvAM='
+			})
+			.reply(200, 'fullname=Jamal%20Fanaian&oauth_token=foo&oauth_token_secret=bar&user_nsid=21207597%40N07&username=jamalfanaian');
```
Comment (alex-seville) [inline test/services.oauth.js]: poor jamal

Comment (jeremyruppel) [inline test/services.oauth.js]: That's what you get for writing the original docs!

File: test/services.oauth.js (hunk 3 above)
Comment (jeremyruppel) [inline test/services.oauth.js]: I straight up could not get lolex to work. Also, the only proper choice of mock timestamp is obviously the event of the world's first temporal displacement, so I fixed that too.

**Pull request merged**